zero/svdknn.py

```diff
@@ -0,0 +1,126 @@
+"""
+Mangaki sparse SVD.
+Author: Jill-Jênn Vie, 2020
+"""
+import numpy as np
+from scipy.sparse import csr_matrix, diags
+from scipy.sparse.linalg import svds
+from sklearn.neighbors import NearestNeighbors
+from zero.recommendation_algorithm import (RecommendationAlgorithm,
+                                           register_algorithm)
+
+
+def remove_mean(sp_matrix, axis=1):
+    '''
+    For each row (resp. column if axis is 0) of a sparse matrix,
+    remove the mean of nonzero elements (resp. the mean of the column)
+    from the nonzero elements.
+    '''
+    mask = sp_matrix.copy()
+    mask.data = np.ones_like(sp_matrix.data)
+    line_sums = sp_matrix.sum(axis=axis).A1
+    if axis == 1:
+        line_counts = mask.sum(axis=axis).A1
+        line_counts[line_counts == 0] = 1
+    else:
+        nb_users, nb_works = sp_matrix.shape
+        line_counts = nb_users * np.ones(nb_works)
+    means = line_sums / line_counts
+    if axis == 0:  # Remove from columns
+        shifted = sp_matrix - mask * diags(means)
+    else:  # Remove from rows
+        shifted = sp_matrix - diags(means) * mask
+    return shifted, means
+
+
+@register_algorithm('svdknn')
+class MangakiSVDKNN(RecommendationAlgorithm):
+    '''
+    Implementation of SVD with sparse matrices.
+    It does not compute the whole matrix for recommendations
+    but the production must be able to do sparse matrix
+    operations effectively.
+    It is 7x faster than svd1.py, and it only relies on numpy/scipy.
```

Comment:
and sklearn :P

Reply:
En fait c'est un commentaire que j'avais copié de SVD :X